.jenkins/infrastructure/docker/build_mystikos_docker_images.Jenkinsfile

```diff
@@ -0,0 +1,139 @@
+    }
+    stages {
+        stage("Checkout") {
+            steps {
+                cleanWs()
+                checkout([$class: 'GitSCM',
+                    branches: [[name: BRANCH_NAME]],
+                    extensions: [],
+                    userRemoteConfigs: [[url: "https://github.com/${params.REPOSITORY_NAME}"]]])
+            }
+        }
+        stage("Base Image") {
+            stages {
+                stage('Build base container') {
+                    steps {
+                        dir(env.BASE_DOCKERFILE_DIR) {
+                            sh """
+                                chmod +x ./build.sh
+                                mkdir build
+                                cd build
+                                ../build.sh -m "${params.MYST_VERSION}" -o "${params.OE_BASE_CONTAINER_TAG}" -u "18.04" -t "${params.MYST_BASE_DOCKER_TAG}"
+                                ../build.sh -m "${params.MYST_VERSION}" -o "${params.OE_BASE_CONTAINER_TAG}" -u "20.04" -t "${params.MYST_BASE_DOCKER_TAG}"
+                            """
+                        }
+                    }
+                }
+                stage('Push base containers to internal repository') {
+                    when {
+                        expression { return params.PUBLISH_CONTAINERS }
+                    }
+                    steps {
+                        script {
+                            tag_bionic = "mystikos-bionic:${params.MYST_BASE_DOCKER_TAG}"
+                            tag_focal = "mystikos-focal:${params.MYST_BASE_DOCKER_TAG}"
+                            docker.withRegistry(params.INTERNAL_REPO, env.INTERNAL_REPO_CREDS) {
+                                base_1804_image = docker.image(tag_bionic)
+                                base_2004_image = docker.image(tag_focal)
+
+                                base_1804_image.push()
+                                base_2004_image.push()
+                                if ( params.TAG_LATEST ) {
+                                    base_1804_image.push('latest')
+                                    base_2004_image.push('latest')
+                                }
+                            }
+                            sh "docker logout"
+                        }
+                    }
+                }
+                stage('Build container requirements') {
+                    steps {
+                        script {
+                            containers_reqs.each { container -> 
+                                try {
+                                    sh """#!/bin/bash
+                                       cd ${WORKSPACE}/${container.value.dir} && make ${container.value.make}
+                                       """
+                                } catch (Exception e) { sh "echo 'Failed: ${container.key}'"}
+                            }
+                        }
+                    }
+                }
+                stage('Build test containers') {
```

Comment:
Perhaps combine these the stages `Build container requirements`, `build test containers`, and `Push test containers to internal repository` so that if a container requirement fails then it would skip the container build.

Alternatively, remove the failing container from the `containers_to_build` list if it fails so that the next stage would not run it. However this may result in subsequent stages / build succeeding even if all containers failed to build.

Reply:
Not all containers have prereqs, but yes, I like the idea of combining the build and push into the same stage!